frontend: label change

NicolaLS opened the pull request:
Label utxos that are to change addresses from our account (subaccounts
respectively) as change to give the user more information about the utxo
when selecting it in coin control.
![requestly-isChange-for-all-tx-small png](https://github.com/BitBoxSwiss/bitbox-wallet-app/assets/90551797/37c871d9-5957-4433-bd61-254b121aceb3)
![requestly-isChange-for-all-tx](https://github.com/BitBoxSwiss/bitbox-wallet-app/assets/90551797/72d5533c-81d7-4d05-96d1-d9f0c64649e7)

thisconnect commented on frontends/web/src/locales/en/app.json:
```diff
@@ -1530,6 +1530,7 @@
     "coincontrol": {
       "address": "Address",
       "addressReused": "Address re-used",
+      "change": "Change",
```
hm not sure if this needs to be translated or if we just keep the label as "change", but no strong opinion.

Beerosagos replied:
I think keeping a label makes sense, in case we want to translate. We can always keep the label untranslated across languages.

thisconnect replied:
the other day we had an error something about WC namespace which we found out it makes more sense to not translate. But translators might miss that and translate anyway.

thisconnect replied:
asked a bit around, people would like to see it translated ie. "Wechselgeld" in German.

lets keep as is.

Beerosagos reviewed (commented):
LGTM! Only one doubt about the first commit comment:

`Possible false negative in case the user used another
wallet and derived some change address behind the current change gap
limit.`

I think that in case the user received on a change address uncovered by our gap limit he will miss the whole utxo, not only the `change` badge. wdyt?

benma commented on backend/coins/btc/account.go:
```diff
@@ -873,12 +874,19 @@ func (account *Account) SpendableOutputs() []*SpendableOutput {
 	}
 	for outPoint, txOut := range utxos {
+		isChange := false
+		for _, subacc := range account.subaccounts {
```
this is the same code as https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/81b3b257f9bad89ec94b1e6939461b2fc6960cc4/backend/coins/btc/account.go#L722-L726

would be nice to move that to a method on Account, with a unit test, so it can be reused

NicolaLS replied:
Maybe it would be better to make the [AccountAddress](https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/160d1adebbabd220d83dfcffbfa0a3dda85b7a9b/backend/coins/btc/addresses/address.go#L34) _change aware_ and then use [getAddress](https://github.com/BitBoxSwiss/bitbox-wallet-app/blob/160d1adebbabd220d83dfcffbfa0a3dda85b7a9b/backend/coins/btc/transaction.go#L218) instead of `IsChange`, here is how it could look like:
https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/3b7fa1246e4cc722a908e0cc9eefedc680000dff

I think it makes sense for AccounAdress to know whether it is change or not. Often times we'll want to get some address anyways, and even if we really only care for the `IsChange` value the overhead (looking at the receiveAdresses) does not matter at all since a hash map is used for the lookup (O(1)) I think this would make the code easier to understand. 

Please lmk what you think @benma

NicolaLS replied:
(added three commits after the initial backend commit, can squash them in case we want to do it like this)

thisconnect replied:
please squash

NicolaLS commented:
> LGTM! Only one doubt about the first commit comment:
> 
> `Possible false negative in case the user used another wallet and derived some change address behind the current change gap limit.`
> 
> I think that in case the user received on a change address uncovered by our gap limit he will miss the whole utxo, not only the `change` badge. wdyt?

Right, I don't know why I thought he would see the utxo. I'll amend the commit message, thanks.

thisconnect reviewed (approved):
frontend LGTM

NicolaLS commented on frontends/web/src/routes/account/send/utxos.tsx:
```diff
@@ -141,6 +141,12 @@ export const UTXOs = ({
                           </Badge> :
                           null
                         }
```
good idea
![image](https://github.com/BitBoxSwiss/bitbox-wallet-app/assets/90551797/4fa35b0d-0df4-47fd-a2b7-93ef0f147829)

thisconnect reviewed (approved):
frontend LGTM

Beerosagos commented:
@NicolaLS is this ready for BE review or does it need other changes?

NicolaLS commented:
> @NicolaLS is this ready for BE review or does it need other changes?

It's ready, I added the `IsChange` method to the account and added a test (also fixed test package name). I still think something like [this](https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/3b7fa1246e4cc722a908e0cc9eefedc680000dff#diff-f35c451e2c99215268316520d28911334d439c389500338e7df38710a3f8e5a7) would be better rather than adding a method to account, but this could be a future PR/refactor too.

Beerosagos commented on backend/coins/btc/account_test.go:
```diff
@@ -13,7 +13,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package btc_test
```
@benma I guess there was a reason why this package was named `btc_test`, but I'm not sure which was it. Is it ok to move inside `btc` package?

NicolaLS replied:
@benma could you take a look at this ? I can't imagine any reason why tests that test code from a certain package should be outside of that package (unit tests). but if there is/was a reason I can just undo the change in this PR.

I already addressed the other change requests from @Beerosagos

benma replied:
What was the reason to change it in the first place?

Often one wants to test the API of a package, not its internals, which is why the tests are often in a separate package (`..._test`). 

It does not really matter much though, so feel free to change it. Afaik the Go stdlib tests are mostly not in external packages either, and that should be a good reference.

Beerosagos commented on backend/coins/btc/account.go:
```diff
@@ -907,6 +902,17 @@ func (account *Account) VerifyExtendedPublicKey(signingConfigIndex int) (bool, e
 }
 
+// IsChange returns true if there is an address corresponding to the provided scriptHashHex in our
+// accounts change address chain. It returns false if no addresss can be found.
```
typo: addresss

Beerosagos commented on backend/coins/btc/account_test.go:
```diff
@@ -190,11 +189,26 @@ func TestSignAddress(t *testing.T) {
 
 }
+
+func TestIsChange(t *testing.T) {
```
Would be nice also to test for false positives: non-change addresses should not be mistakenly identified as change

NicolaLS commented:
@Beerosagos adressed all change requests and @benma is okay with changing the tests package if I understood correctly:
 https://github.com/BitBoxSwiss/bitbox-wallet-app/pull/2788#discussion_r1770985912

So should be ready for final review/merge.

Beerosagos reviewed (approved):
tACK :pray:

thisconnect commented:
please add a changelog entry 🙏

NicolaLS commented:
@Beerosagos can we merge this?

Beerosagos commented:
> @Beerosagos can we merge this?

Yes, thanks

NicolaLS commented:
lint fail should be: 
https://github.com/BitBoxSwiss/bitbox-wallet-app/commit/cc9fb570194ea55f427330dcb27c3923b75c0ad1

trivy also intermittent fail, @thisconnect could you run CI again and if it succeeds merge the PR?

thisconnect commented:
@NicolaLS please rebase

Beerosagos commented:
> @NicolaLS please rebase

Sorry, I missed your comment before merging :pray:
